pkg/controllers/management/clusterupstreamrefresher/cluster_upstream_refresher.go

```diff
@@ -0,0 +1,274 @@
+		}
+		cluster = cluster.DeepCopy()
+		apimgmtv3.ClusterConditionUpdated.False(cluster)
+		apimgmtv3.ClusterConditionUpdated.Message(cluster, fmt.Sprintf("[Syncing error%s] %s", syncFailed, err.Error()))
+
+		cluster, statusErr = c.clusterClient.Update(cluster)
+		if statusErr != nil {
+			return cluster, statusErr
+		}
+
+		if upstreamConfig == nil {
+			return cluster, err
+		}
+	} else if strings.Contains(apimgmtv3.ClusterConditionUpdated.GetMessage(cluster), "[Syncing error") {
+		cluster = cluster.DeepCopy()
+		apimgmtv3.ClusterConditionUpdated.True(cluster)
+		apimgmtv3.ClusterConditionUpdated.Message(cluster, "")
+		cluster, err = c.clusterClient.Update(cluster)
+		if err != nil {
+			return cluster, err
+		}
+	}
+
+	var initialClusterConfig, appliedClusterConfig, upstreamClusterConfig, upstreamSpec interface{}
+	// for other cloud drivers, please edit HERE
+	switch cloudDriver {
+	case apimgmtv3.ClusterDriverEKS:
+		initialClusterConfig = cluster.Spec.EKSConfig
+		appliedClusterConfig = cluster.Status.AppliedSpec.EKSConfig
+		upstreamClusterConfig = cluster.Status.EKSStatus.UpstreamSpec
+		upstreamSpec = upstreamConfig.eksConfig
+	}
+
+	// compare saved cluster.Status...UpstreamSpec with upstreamSpec,
+	// if there is difference then update cluster.Status...UpstreamSpec
+	if !reflect.DeepEqual(upstreamClusterConfig, upstreamSpec) {
+		logrus.Infof("updating cluster [%s], upstream change detected", cluster.Name)
+		cluster = cluster.DeepCopy()
+		// for other cloud drivers, please edit HERE
+		switch cloudDriver {
+		case apimgmtv3.ClusterDriverEKS:
+			cluster.Status.EKSStatus.UpstreamSpec = upstreamConfig.eksConfig
+		}
+		cluster, err = c.clusterClient.Update(cluster)
+		if err != nil {
+			return cluster, err
+		}
+	}
+
+	// check if cluster is still updating changes
+	if !reflect.DeepEqual(initialClusterConfig, appliedClusterConfig) {
+		logrus.Infof("cluster [%s] currently updating, skipping spec sync", cluster.Name)
+		return cluster, nil
+	}
+
+	// check for changes between upstream spec on cluster and initial ClusterConfig object
+	specMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(initialClusterConfig)
+	if err != nil {
+		return cluster, err
+	}
+
+	upstreamSpecMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(upstreamSpec)
+	if err != nil {
+		return cluster, err
+	}
+
+	var updateClusterConfig bool
+	for key, value := range upstreamSpecMap {
+		if specMap[key] == nil {
+			continue
+		}
+		if reflect.DeepEqual(specMap[key], value) {
+			continue
+		}
+		updateClusterConfig = true
+		specMap[key] = value
+	}
+
+	if !updateClusterConfig {
+		logrus.Infof("cluster [%s] matches upstream, skipping spec sync", cluster.Name)
+		return cluster, nil
+	}
+
+	// for other cloud drivers, please edit HERE
+	switch cloudDriver {
+	case apimgmtv3.ClusterDriverEKS:
+		err = runtime.DefaultUnstructuredConverter.FromUnstructured(specMap, cluster.Spec.EKSConfig)
+	}
+	if err != nil {
+		return cluster, err
+	}
```

Comment:
Can the error be checked where it's created? This looks like it could get into the same issue @thedadams found previously.